File: app/src/main/java/allgoritm/com/centrifuge/CompositeDisposablesMap.java
package allgoritm.com.centrifuge;

import io.reactivex.disposables.Disposable;

import java.util.HashMap;
import java.util.Map;


public class CompositeDisposablesMap {

    private final Map<String, Disposable> subscriptions;

    public CompositeDisposablesMap() {
        subscriptions = new HashMap<>();
    }

    public synchronized void put(String key, Disposable subscription) {
        clear(key);
        subscriptions.put(key, subscription);
    }

    public synchronized boolean isUnsubscribed(String key) {
        return subscriptions.get(key).isDisposed();
    }

    public synchronized void clear(String key) {
        if (subscriptions.containsKey(key)) {
            clearSubscription(subscriptions.get(key));
            subscriptions.remove(key);
        }
    }

    public synchronized boolean containsKey(String key) {
        return subscriptions.containsKey(key);

    }

    private synchronized void clearSubscription(Disposable subscription) {
        if (subscription != null && !subscription.isDisposed()) {
            subscription.dispose();
        }
    }

    public synchronized void clearAll() {
        for (Map.Entry<String, Disposable> entry : subscriptions.entrySet()) {
            clearSubscription(entry.getValue());
        }
        subscriptions.clear();
    }

}

File: centrifuge/src/main/java/allgoritm/com/centrifuge/v1/util/CompositeDisposablesMap.kt
package allgoritm.com.centrifuge.v1.util

import io.reactivex.disposables.Disposable
import java.util.*


internal class CompositeDisposablesMap {

    private val subscriptions: MutableMap<String, Disposable>

    init {
        subscriptions = HashMap()
    }

    @Synchronized
    fun put(key: String, subscription: Disposable) {
        clear(key)
        subscriptions[key] = subscription
    }

    @Synchronized
    fun isUnsubscribed(key: String): Boolean {
        return subscriptions[key]!!.isDisposed
    }

    @Synchronized
    fun clear(key: String) {
        if (subscriptions.containsKey(key)) {
            clearSubscription(subscriptions[key])
            subscriptions.remove(key)
        }
    }

    @Synchronized
    fun containsKey(key: String): Boolean {
        return subscriptions.containsKey(key)

    }

    @Synchronized
    private fun clearSubscription(subscription: Disposable?) {
        if (subscription != null && !subscription.isDisposed) {
            subscription.dispose()
        }
    }

    @Synchronized
    fun clearAll() {
        for ((_, value) in subscriptions) {
            clearSubscription(value)
        }
        subscriptions.clear()
    }

}
